src/tribler/core/components/ipv8/eva/transfer/base.py

```diff
@@ -57,40 +93,71 @@ def finish(self, *, result: Optional[TransferResult] = None, exception: Optional
 
             # To prevent "Future exception was never retrieved" error when the future is not used
             self.future.exception()
-
-            if self.on_error:
+            self.parent_task_group.add(
                 asyncio.create_task(self.on_error(self.peer, exception))
+            )
 
         if result:
             self.logger.debug(f'Finish with result: {result}')
             self.future.set_result(result)
-            if self.on_complete:
+            self.parent_task_group.add(
                 asyncio.create_task(self.on_complete(result))
+            )
 
-        self._release()
+        await self._release()
 
     def on_future_cancelled(self, _):
         if not self.future.cancelled():
             return
 
-        if self.on_error:
-            self.logger.warning('Future was cancelled')
-            exception = TransferCancelledException('The future was cancelled', self)
-            asyncio.create_task(self.on_error(self.peer, exception))
-
-        self._release()
-
-    async def terminate_by_timeout_task(self):
-        timeout = self.settings.timeout_interval_in_sec
-        remaining_time = timeout
-
-        while self.settings.terminate_by_timeout_enabled:
+        self.logger.warning('Future was cancelled')
+        exception = TransferCancelledException('The future was cancelled', self)
+        self.parent_task_group.add(
+            [
+                asyncio.create_task(self.on_error(self.peer, exception)),
+                asyncio.create_task(self._release())
```

Comment:
**nitpick**: Currently, the `on_error` callback and the `_release` coroutine are started as two parallel tasks, so it is impossible to say in what order they will be executed (it depends on the loop scheduler implementation). The same applies to the `finish` method code - the order of `self.on_error/self.on_complete` and `await self._release()` is not guaranteed. The more predictable way is to schedule a new helper task that calls the success/error callback and the `_release` method in a predictable order:

```python
def final_task(self, callback_coroutine):
    await callback_coroutine
    await self._release()
```

Reply:
It was fixed by making `self._release()` synchronous.